core/templates/dev/head/services/assets-backend-api.service.spec.ts

```diff
@@ -26,240 +26,516 @@ import { ImageFileObjectFactory } from
       var failHandler = jasmine.createSpy('fail');
 
-      var requestUrl = UrlInterpolationService.interpolateUrl(
-        'assetsdevhandler/exploration/<exploration_id>/assets/image/<filename>',
-        {
-          exploration_id: '0',
-          filename: 'myfile.png'
+      spyOn(window, 'Blob').and.callFake(function() {
+        throw new TypeError();
+      });
+
+      window.BlobBuilder = function BlobBuilder() {
+        this.blob = new Object({ blobBuilder: true });
+        this.append = function append(data) {};
+        this.getBlob = function getBlob(contentType) {
+          return this.blob;
+        };
+      };
+
+      $httpBackend.expect('GET', audioRequestUrl).respond(201, 'Error');
+      expect(AssetsBackendApiService.isCached('myfile.mp3')).toBe(false);
+
+      AssetsBackendApiService.loadAudio('0', 'myfile.mp3').then(
+        successHandler, failHandler);
+      $httpBackend.flush();
+      expect(successHandler).toHaveBeenCalledWith(
+        audioFileObjectFactory.createNew(
+          'myfile.mp3',
+          { blobBuilder: true }
+        ));
+      expect(failHandler).not.toHaveBeenCalled();
+      $httpBackend.verifyNoOutstandingExpectation();
+    });
+
+    it('should successfully save an audio', function(done) {
+      var successMessage = 'Audio was successfully saved.';
+      // @ts-ignore in order to ignore JQuery properties that should
+      // be declarated.
+      spyOn($, 'ajax').and.callFake(function() {
+        var d = $.Deferred();
+        d.resolve(successMessage);
+        return d.promise();
+      });
+
+      AssetsBackendApiService.saveAudio('0', 'a.mp3', new File([], 'a.mp3'))
+        .then(function(response) {
+          expect(response).toBe(successMessage);
+        }).then(done, done.fail);
+
+      // $q Promises need to be forcibly resolved through a JavaScript digest,
+      // which is what $apply helps kick-start.
+      $rootScope.$apply();
+    });
+
+    it('should handle rejection when saving a file fails', function(done) {
+      var errorMessage = 'Error on saving audio';
+      // @ts-ignore in order to ignore JQuery properties that should
+      // be declarated.
+      spyOn($, 'ajax').and.callFake(function() {
+        var d = $.Deferred();
+        d.reject({
+          /* eslint-disable quotes */
+          responseText: ")]}'\n{ \"message\": \"" + errorMessage + "\" }"
+          /* eslint-enable quotes */
         });
+        return d.promise();
+      });
 
-      $httpBackend.expect('GET', requestUrl).respond(201, 'image data');
+      AssetsBackendApiService.saveAudio('0', 'a.mp3', new File([], 'a.mp3'))
+        .then(done, function(response) {
+          expect(response).toEqual({
+            message: errorMessage
+          });
+          done();
+        });
+
+      // $q Promises need to be forcibly resolved through a JavaScript digest,
+      // which is what $apply helps kick-start.
+      $rootScope.$apply();
```

Comment:
is this related to #8521? If yes please add a TODO comment here so that we can remove it while resolving #8521.

Reply:
I'm really not sure. I'm using it just because saveAudio` calls `$q`. #8521 is about this case too? @bansalnitish

Reply:
No, it's not related to that.